More complete tests and better type safety.

geon opened the pull request:
I noticed the `type` property returned by `imageSize` was optional and typed as a string. Now it is required and typed as a string union of all supported image formats.

geon committed:
Refactor. Check type directly.

geon committed:
Refactor. Changed typeHandlers to an object. Improves type safety.

geon committed:
Refactor. Removed pointless check. `size` is always defined.

geon committed:
Refactor. Inverted condition for Early Return.

geon committed:
Refactor. Added type IconType and asserted it.

geon committed:
Refactor. Types should be CamelCase.

geon committed:
Refactor. Added parsing of ImageType.

geon committed:
Refactor. Added type ImageFormat and parsing.

geon committed:
Added image types to valid spec.

geon committed:
Added valid spec.

geon committed:
Better test.

geon committed:
Removed bad data from spec.

geon committed:
Better test.

geon committed:
Better test error message.

geon committed:
Added orientations to valid spec.

geon committed:
Better test.

geon committed:
Added missing files to valid spec.

geon committed:
Better test.

geon committed:
Better test.

geon committed:
Refactor. Removed unused default size.

geon committed:
Refactor. Removed pointless cast.

geon committed:
Made IImageCalculationResult.type required.

geon committed:
Refactor. Removed unused import.

geon committed:
Refactor. Added separate type for the individual IImage calculation results.

geon committed:
Refactor. Made the relationship between the types clearer.

geon committed:
Updated contributors.

geon commented:
SonarCloud complains about repetition in the valid spec file. The spec data could easily be moved to a json file if you care about that, or even better; replaced with a snapshot test.

geon commented:
Anything I can do to get this merged? Split it up? Remove something?

netroy commented:
Thanks for putting all this effort to improve this project. 
There is definitely stuff in here that I'd like to merge, but there is also stuff that reverts some of the changes I intentionally made, like switching to Maps for lookup.

I appreciate this, but in the future (for any project) please 
1. check with the maintainers about any refactoring that they are willing to accept, before you start the refactoring.
2. avoid creating large refactor PRs, unless there is an absolute need for it. smaller, focused PRs that address specific issues are much easier to review and merge. 
3. avoid changing code styles, casing, etc. since the original code is written to the style preferred by the people who'll have to maintain the project in the long run

I'll most likely have to sieve through the changes that make sense to me, to extract them out into separate commits (with attribution to you, of course).

However, I don't see anything in here that needs to be addressed urgently, so it might take a while, since I'm currently occupied with a couple of other projects.

In the future, if you'd like to discuss your ideas for this project, feel free to open an issue first, so that we can first align on how to address it.

Again, thanks for doing this. I'll try to get to this as soon as I can.

geon commented:
Sorry for my late reply!

> switching to Maps for lookup
Yes. I specifically noted that you had changed that in a recent commit. However, an object gives much better typing:

    const obj = { foo: 123, bar: "text" } as const;
    const val = obj["foo"];
    // typeof val is 123

Compare it to maps:

    const map = new  Map([["foo", 123], ["bar", "text"]]);
    const val = map.get("foo");
    // typeof val is number | string | undefined

You simply can't get the same level of type safety with maps. Map is good for performance when inserting and removing a lot, when you need non-string keys, or dynamic input keys, since you don't risk conflicts with prototypes etc. None of that is relevant in this library. Was there any other reason to use Map?

1: Sorry about that. I didn't mean to step on any toes. I just started the refactoring out of curiosity to see if I could get the typing working.

2: I wanted to solve the exposed typing. It required a lot of changes, most of which don't make much sense individually. Possibly, the test updates could be separated to it's own pull request. Would you like me to do it?

3: Noted. Sorry about that.

geon commented:
Everything up to and including 7b587fe3bfb1378843e31042c1f9b8ef1901dec0 is just

* Added test data
* Explicit test data instead of the default value
* Removed invalid test data that were never used, because the data being tested was missing them
* Refactored test code.

After those commits, the test data is complete and correct, and the test is simplified to a single deepStrictEqual.

I included all commits separately to show in clear small steps that nothing has been broken or removed.

geon commented:
Feel free to comment on each commit, and I can update/split the pull request.
